benchmarks/struct_indices/spider/evaluate.py

```diff
@@ -0,0 +1,291 @@
+    )
+    parser.add_argument(
+        "--predict-dir",
+        type=str,
+        required=True,
+        help="Path to the directory of generated SQL files.",
+    )
+    parser.add_argument(
+        "--model",
+        type=str,
+        default="gpt-4",
+        choices=["gpt-4", "gpt-3.5-turbo"],
+        help="The model used to perform evaluation.",
+    )
+    parser.add_argument(
+        "--use-cache",
+        action="store_true",
+        help="Whether to use the cached results or not.",
+    )
+    args = parser.parse_args()
+
+    # Create the LlamaIndexes for all databases.
+    if args.model in ["gpt-3.5-turbo", "gpt-4"]:
+        llm = ChatOpenAI(model=args.model, temperature=0)
+    else:
+        llm = OpenAI(model=args.model, temperature=0)
+
+    # Load all examples.
+    train, dev = load_examples(args.spider_dir)
+
+    # Load all generated SQL queries.
+    with open(os.path.join(args.predict_dir, "train_pred.sql"), "r") as f:
+        train_pred_sqls = f.readlines()
+    with open(os.path.join(args.predict_dir, "dev_pred.sql"), "r") as f:
+        dev_pred_sqls = f.readlines()
+
+    # Load all gold SQL queries and database names.
+    train_dbs = []
+    dev_dbs = []
+    train_gold_sqls = []
+    dev_gold_sqls = []
+    with open(os.path.join(args.spider_dir, "train_gold.sql"), "r") as f:
+        for line in f.readlines():
+            line = line.strip().split("\t")
+            train_gold_sqls.append(line[0])
+            train_dbs.append(line[1])
+    with open(os.path.join(args.spider_dir, "dev_gold.sql"), "r") as f:
+        for line in f.readlines():
+            line = line.strip().split("\t")
+            dev_gold_sqls.append(line[0])
+            dev_dbs.append(line[1])
+
+    # Create Llama indexes on the databases.
+    indexes = create_indexes(spider_dir=args.spider_dir, llm=llm, temperature=0.0)
+
+    # Run SQL queries on the indexes and get NL answers.
+    train_pred_results = _get_answers(
+        llm,
+        indexes,
+        train_dbs,
+        train_pred_sqls,
+        train,
+        os.path.join(args.predict_dir, "train_pred_results.json"),
+        args.use_cache,
+    )
+    train_gold_results = _get_answers(
+        llm,
+        indexes,
+        train_dbs,
+        train_gold_sqls,
+        train,
+        os.path.join(args.predict_dir, "train_gold_results.json"),
+        args.use_cache,
+    )
+    dev_pred_results = _get_answers(
+        llm,
+        indexes,
+        dev_dbs,
+        dev_pred_sqls,
+        dev,
+        os.path.join(args.predict_dir, "dev_pred_results.json"),
+        args.use_cache,
+    )
+    dev_gold_results = _get_answers(
+        llm,
+        indexes,
+        dev_dbs,
+        dev_gold_sqls,
+        dev,
+        os.path.join(args.predict_dir, "dev_gold_results.json"),
+        args.use_cache,
+    )
+
+    # Evaluate results.
+    train_match = _match_answers(
+        llm,
```

Comment:
curious: from blog, my understanding is that you are using ChatGPT to evaluate results from different LLMs.

But here, looks like the same LLM is used for text-to-sql and for evaluation. Curious if I missed anything.

Reply:
The SQL generation is done separately using `generate_sql.py`, for which I used different GPT models. There is a command line argument in `generate_sql.py` that allows for different LLMs to be used. Hope this clarifies.

Reply:
In this file, the index is called with `mode="sql"` so it is only executing the generated SQL queries.

Reply:
Got it, thanks for the clarification!